src/test/clojure/xtdb/trie_test.clj

```diff
@@ -136,12 +137,286 @@
+                    (for [[page min-vf max-vt] [[0 micros-2018 micros-2019]
+                                                [1 micros-2019 micros-2020]
+                                                [2 micros-2020 micros-2021]
+                                                [3 micros-2021 micros-2022]
+                                                [4 micros-2020 Long/MAX_VALUE]]]
+                      (->mock-merge-plan-page page min-vf max-vt))
+                    query-bounds)
+                   (mapv #(scan/load-page % nil nil))))]
+
+
+      (let [query-bounds (TemporalBounds.)]
+        ;; stricly inside 2020
+        (.lte (.getSystemFrom query-bounds) (inc at-time))
+        (.gt (.getSystemTo query-bounds) (inc at-time))
+        (.lte (.getValidFrom query-bounds) (dec to-time))
+        (.gte (.getValidTo query-bounds) (inc at-time))
+
+        (t/is (= [2 4] (query-bounds->pages query-bounds))))
+
+      (let [query-bounds (TemporalBounds.)]
+        ;; 2020
+        (.lte (.getSystemFrom query-bounds) at-time)
+        (.gt (.getSystemTo query-bounds) at-time)
+        (.lte (.getValidFrom query-bounds) to-time)
+        (.gte (.getValidTo query-bounds) at-time)
+
+        (t/is (= [1 2 4] (query-bounds->pages query-bounds))))
+
+      (let [query-bounds (TemporalBounds.)]
+        ;; no system-bounds
+        (.lte (.getValidFrom query-bounds) to-time)
+        (.gte (.getValidTo query-bounds) at-time)
+
+        (t/is (= [1 2 3 4] (query-bounds->pages query-bounds))))
+
+      (let [query-bounds (TemporalBounds.)]
+        ;; no bounds
+
+        (t/is (= [0 1 2 3 4] (query-bounds->pages query-bounds)))))))
+
+(defn ->constantly-int-pred [v]
+  (reify IntPredicate (test [_ _] v)))
+
+(defn ->constantly-pred [v]
+  (reify Predicate (test [_ _] v)))
+
+(defn ->merge-tasks [segments query-bounds]
+  (->> (HashTrieKt/toMergePlan segments (->constantly-pred true) query-bounds)
+       (into [] (keep (fn [^MergePlanTask mpt]
+                        (when-let [leaves (scan/->merge-task
+                                           (for [^MergePlanNode mpn (.getMpNodes mpt)
+                                                 :let [{:keys [seg page-bounds-fn]} (.getSegment mpn)
+                                                       node (.getNode mpn)]]
+                                             (reify MergePlanPage
+                                               (load-page [_ _ _] {:seg seg :page-idx (.getDataPageIndex ^ArrowHashTrie$Leaf node)})
+                                               (test-metadata [_] true)
+                                               (temporal-bounds [_] (page-bounds-fn (.getDataPageIndex ^ArrowHashTrie$Leaf node)))))
+                                           query-bounds)]
+                          {:path (vec (.getPath mpt))
+                           :pages (mapv #(scan/load-page % nil nil) leaves)}))))
+       (sort-by :path)))
+
+(deftest test-to-merge-tasks-with-temporal-filters
+  (let [
```

Comment:
This is a test I had before the refactoring. It essentially combines `toMergePlan` with `->merge-task`. If we consider the above test (`test-->merge-task`) sufficient I can also delete the code related to this.